## Data Science Project
#### Riley So

#### Introduction
The Computer Science department at the University of British Columbia (UBC) is running a project that uses a Minecraft server to study players behaviors. Actions performed by players are logged, creating a data set that captures behavior, characteristiscs, and engagement patterns.

However running this type of research project takes time and money. The team needs to buy software licenses, manage server capacity, and find the right players to join. One way to reach players and increase player count is through a game-related newsletter. Knowing which players are most likely to subscribe can help the team focus their recruitment and use their resources more wisely.

##### **Question**
In order to help the research team target their recruitment efforts, we aim to answer the following question.
> Can player characteristics such as age, gender, experience level, and total hours played predict whether a player subscribes to a game-related newsletter?

#### Data Description

There are two datasets:
- `players.csv`: Contains demographic and self-reported experience information for 196 players.
- `sessions.csv`: Contains 1,535 records of individual gameplay sessions.

To explore this data we will begin by loading the data.


In [5]:
library(tidyverse)

# Load players and sessions data
players <- read_csv("data/players.csv")
sessions <- read_csv("data/sessions.csv")

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Now we will provide a find the dimensions of the data, its structure, and its variable types.

In [9]:
#Data Set Dimensions
players_dim <- players |> dim()
sessions_dim <- sessions |> dim()

players_dim
sessions_dim

#Structure and Variable Types
players |> glimpse()
sessions |> glimpse()

[1] 196   7

[1] 1535    5

Rows: 196
Columns: 7
$ experience   <chr> "Pro", "Veteran", "Veteran", "Amateur", "Regular", "Amate…
$ subscribe    <lgl> TRUE, TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, FALSE, TRUE, T…
$ hashedEmail  <chr> "f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8…
$ played_hours <dbl> 30.3, 3.8, 0.0, 0.7, 0.1, 0.0, 0.0, 0.0, 0.1, 0.0, 1.6, 0…
$ name         <chr> "Morgan", "Christian", "Blake", "Flora", "Kylie", "Adrian…
$ gender       <chr> "Male", "Male", "Male", "Female", "Male", "Female", "Fema…
$ Age          <dbl> 9, 17, 17, 21, 21, 17, 19, 21, 17, 22, 23, 17, 25, 22, 17…
Rows: 1,535
Columns: 5
$ hashedEmail         <chr> "bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8a…
$ start_time          <chr> "30/06/2024 18:12", "17/06/2024 23:33", "25/07/202…
$ end_time            <chr> "30/06/2024 18:24", "17/06/2024 23:46", "25/07/202…
$ original_start_time <dbl> 1.71977e+12, 1.71867e+12, 1.72193e+12, 1.72188e+12…
$ original_end_time   <dbl> 1.71977e+12, 1.71867e+12, 1.72193e+12, 1.72188e+

The `players.csv` dataset contains 196 rows and 7 columns, while `sessions.csv` has 1,535 rows and 5 columns. Each row in `players.csv` represents a unique player, while each row in `sessions.csv` represents one play session. The `players` dataset includes categorical, numeric, and logical types. `sessions` is mainly timestamp-based with a unique ID (`hashedEmail`) used in both datasets.

Before summarising statistics we want to account for missing values.

In [10]:
players |> summarise(across(everything(), ~ sum(is.na(.))))
sessions |> summarise(across(everything(), ~ sum(is.na(.))))

experience subscribe hashedEmail played_hours name gender Age
1 0          0         0           0            0    0      2

hashedEmail start_time end_time original_start_time original_end_time
1 0           0          2        0                   2

- `players.csv` has **2 missing values** in the `Age` column.
- `sessions.csv` has **2 missing values** in `end_time` and `original_end_time`.
No other missing values are present.

In [18]:
players |> summary()
sessions |> summary()

  experience        subscribe       hashedEmail         played_hours    
 Length:196         Mode :logical   Length:196         Min.   :  0.000  
 Class :character   FALSE:52        Class :character   1st Qu.:  0.000  
 Mode  :character   TRUE :144       Mode  :character   Median :  0.100  
                                                       Mean   :  5.846  
                                                       3rd Qu.:  0.600  
                                                       Max.   :223.100  
                                                                        
     name              gender               Age       
 Length:196         Length:196         Min.   : 8.00  
 Class :character   Class :character   1st Qu.:17.00  
 Mode  :character   Mode  :character   Median :19.00  
                                       Mean   :20.52  
                                       3rd Qu.:22.00  
                                       Max.   :50.00  
                               

 hashedEmail         start_time          end_time         original_start_time
 Length:1535        Length:1535        Length:1535        Min.   :1.712e+12  
 Class :character   Class :character   Class :character   1st Qu.:1.716e+12  
 Mode  :character   Mode  :character   Mode  :character   Median :1.719e+12  
                                                          Mean   :1.719e+12  
                                                          3rd Qu.:1.722e+12  
                                                          Max.   :1.727e+12  
                                                                             
 original_end_time  
 Min.   :1.712e+12  
 1st Qu.:1.716e+12  
 Median :1.719e+12  
 Mean   :1.719e+12  
 3rd Qu.:1.722e+12  
 Max.   :1.727e+12  
 NA's   :2          

##### Summary of Data
**players.csv**

- **Age**: Ranges from 8 to 50 years (Mean: 20.5, Median: 19). Most players are in their late teens or early 20s. Two values are missing.
- **played_hours**: Strongly right-skewed — Median is only 0.1, but Max is 223 hours. Most players spent very little time on the server, with a few heavy users.
- **subscribe**: About 73% (144/196) of players subscribed to the newsletter this is our target variable.
- **experience & gender**: Character columns that will be converted to categorical for modeling and grouped visualization.
- **name & hashedEmail**: Identifiers, not used for modeling directly.

**sessions.csv**

- **original_start_time / original_end_time**: Unix timestamps covering mid-2024. Can be used to analyze peak play times or session durations. Two `original_end_time` values are missing.
- **start_time / end_time**: Human-readable time strings. Useful for plotting session timing trends.
- **hashedEmail**: Common key with `players.csv` for potential grouping but not needed for prediction.



Additionally lets look at the **categorical data**

In [20]:
# players
players |> count(experience)
players |> count(gender)
players |> count(subscribe)

# sessions
sessions |> distinct(hashedEmail) |> count(name = "unique_players")

experience n 
1 Amateur    63
2 Beginner   35
3 Pro        14
4 Regular    36
5 Veteran    48

gender            n  
1 Agender             2
2 Female             37
3 Male              124
4 Non-binary         15
5 Other               1
6 Prefer not to say  11
7 Two-Spirited        6

subscribe n  
1 FALSE      52
2  TRUE     144

unique_players
1 125

**Observations of categorical data**
- Most players classify as "Veteran" or "Pro".
- Gender appears evenly distributed.
- Subscription rates are close to balanced.
- 85+ unique players have sessions recorded, indicating repeat activity.

#### **Takeaway** 
Both datasets are clean and contain relevant features.
- `players.csv` is suited for modeling the subscription prediction.
- `sessions.csv` will support additional insights.

### Data Cleaning
Now that we have described and explored the data, lets prepare the data for anlaysis. To do this we will:
- Remove or impute missing values
- Convert characters to factors
- Prepare data for modeling

As seen above we found 2 missing values in the `Age` column. Since this is a small portion of the dataset (2 of 196) we will **remove** these rows.

In [26]:
# Clean Data
players_clean <- players |> filter(!is.na(Age))

experience subscribe
1   Pro         TRUE    
2   Veteran     TRUE    
3   Veteran    FALSE    
4   Amateur     TRUE    
5   Regular     TRUE    
6   Amateur     TRUE    
7   Regular     TRUE    
8   Amateur    FALSE    
9   Amateur     TRUE    
10  Veteran     TRUE    
11  Veteran     TRUE    
12  Pro         TRUE    
13  Beginner    TRUE    
14  Amateur     TRUE    
15  Veteran     TRUE    
16  Amateur    FALSE    
17  Beginner    TRUE    
18  Amateur     TRUE    
19  Amateur     TRUE    
20  Regular     TRUE    
21  Regular     TRUE    
22  Amateur    FALSE    
23  Beginner    TRUE    
24  Beginner    TRUE    
25  Amateur     TRUE    
26  Regular     TRUE    
27  Regular     TRUE    
28  Veteran    FALSE    
29  Amateur     TRUE    
30  Veteran    FALSE    
⋮   ⋮          ⋮        
165 Regular     TRUE    
166 Amateur    FALSE    
167 Beginner   FALSE    
168 Veteran     TRUE    
169 Veteran    FALSE    
170 Pro        FALSE    
171 Beginner   FALSE    
172 Veteran     TRUE    
173 Regular    FALSE    
174 Veteran     TRUE    
175 Amateur     TRUE    
176 Pro        FALSE    
177 Veteran    FALSE    
178 Amateur     TRUE    
179 Veteran     TRUE    
180 Amateur    FALSE    
181 Amateur     TRUE    
182 Pro        FALSE    
183 Amateur     TRUE    
184 Pro        FALSE    
185 Regular    FALSE    
186 Veteran     TRUE    
187 Amateur     TRUE    
188 Beginner    TRUE    
189 Amateur    FALSE    
190 Amateur     TRUE    
191 Amateur     TRUE    
192 Veteran    FALSE    
193 Amateur    FALSE    
194 Amateur    FALSE    
    hashedEmail                                                     
1   f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8b53e192d
2   f3c813577c458ba0dfef80996f8f32c93b6e8af1fa939732842f2312358a88e9
3   b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3c5a9d2118eb7ccbb28
4   23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4fa7a5a659ff443a0eb5
5   7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb0af4d48fcce2420f3e
6   f58aad5996a435f16b0284a3b267f973f9af99e7a89bee0430055a44fa92f977
7   8e594b8953193b26f498db95a508b03c6fe1c24bb5251d392c18a0da9a722807
8   1d2371d8a35c8831034b25bda8764539ab7db0f63938696917c447128a2540dd
9   8b71f4d66a38389b7528bb38ba6eb71157733df7d1740371852a797ae97d82d1
10  bbe2d83de678f519c4b3daa7265e683b4fe2d814077f9094afd11d8f217039ec
11  7a4686586d290c67179275c7c3dfb4ea02f4d317d9ee0e2cee98baa27877a875
12  4caa42e1b20511552434978171dcf7283fb6eb857eb871327e39bfc328002309
13  4bed6f49b672974f706aa7f020b4f29020326bf594520593557cd381a51559b8
14  3caa832978e0596779f4ee7c686c4592fb6de893925025d25bf26be9a8cd03f5
15  6f9acf8ea9956fe817895c78d10e1e25c11aba335a451e5b0322b073ee5008c4
16  2313a06afe47eacc28ff55adf6f072e7d12b0d12d7cbae29656f9029ec72bb53
17  9db37a1d381697f6c83e3b97b276e38bb9d9c811ba4596edb75d4dbfea6415b0
18  a175d4741dc84e6baf77901f6e8e0a06f54809a34e6b5211159bced346f7fb3e
19  ab1f44f93c3b828f55458971db393052d9711df3e0e7ff69540bfebfcec555ff
20  20ae3f8b5323674fe6e537bc91928841cfc2a00c1b36522fd25370bb6fa89d36
21  bc704ff2bc676dbf48ee41b9e11481c1387bf758ad318f2428f336e3fecc6660
22  200db0417ee70863c7e2a0f214c737476fb86de5869d0b907a077d726357e687
23  f9ac013b2f0bc2bd4928a6a0fc8a0aae8b5c4f2670cf9e18f547b7e72034c965
24  89c8e07d6a2aa102d9cd05086d51b2e41195c9c807dd976a06f4262ff36d0cca
25  119f01b9877fc5ea0073d05602a353b91c4b48e4cf02f42bb8d661b46a34b760
26  5baba1651a0b92788bc0d6dcdf00be64af1cf9f0015bbeed204ddabd1d872e83
27  50cec73a632e4be7b155c3df5165de28b2c5baf092abcb7f509f5645776a56c2
28  f8acd26a5e452b682b9f8b20108caef1ee2e745efe08e95806980bd2388a3fea
29  4b01bce3f141289709e8278b02ba5d2aaa7105d7ccb9c7deb37670a80e332774
30  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b13f9e0084b08f8070bd
⋮   ⋮                                                               
165 c121e4d197469bea90e21c0495001f4e21824adb98cbc6a40d54e66617cee174
166 7096b30972a32029244fc6171f21bc5996b51b8d75df7a94f003dead428988f9
167 42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898abcf1177d9822bd39f7
168 9ba67e127e4e70d909f6237644bb4674b17